Add support for Pluggable Discoveries (#1333)
arduino/arduino-cli: docs/pluggable-discovery-specification.md

> Discovery tools are a special kind of tool used to find supported boards, a platform developer can create its own

Reviewer: ```suggestion
Discovery tools are a special kind of tool used to find supported boards, a platform developer can create their own
```
I don't think "its" is appropriate here ([reference](https://en.wikipedia.org/wiki/Gender_neutrality_in_languages_with_gendered_third-person_pronouns#It_and_one_as_gender-neutral_pronouns)).

My suggested replacement is debatable, but I think it's reasonable. See the answer to [this question](https://www.chicagomanualofstyle.org/qanda/data/faq/topics/Pronouns.html#:~:text=the%20singular%20%E2%80%9Cthey%2C%E2%80%9D) for a reasonable take on the subject.

> These tools must be in the form of executables that can be launched as a subprocess using a `platform.txt` command line

Author: I added a small introduction with https://github.com/arduino/arduino-cli/pull/1333/commits/fb1dde149b745e1c684355e870c60c8e4805d461.

> These tools must be in the form of executables that can be launched as a subprocess using a `platform.txt` command line

Reviewer: Looks great!

recipe. They will communicate to the parent process via stdin/stdout, in particular a discovery will accept commands as
plain text strings from stdin and will send answers back in JSON format on stdout. Each tool will implement the commands

> to list and enumerate ports for a specific protocol as specified in this document.

Reviewer: This change was reverted by https://github.com/arduino/arduino-cli/commit/fb1dde149b745e1c684355e870c60c8e4805d461

> to list and enumerate ports for a specific protocol as specified in this document.

Author: Yep, I messed up a bit, will fix right away.

> to list and enumerate ports for a specific protocol as specified in this document.

Author: Fixed: https://github.com/arduino/arduino-cli/commit/a42e2931a5aa3673543969dc27a86ce013524416

> to list and enumerate ports for a specific protocol as specified in this document.

Reviewer: Looks perfect now!

command should be called if the client wants to pause the discovery for a while. The response to the command is:

```JSON
{
  "eventType": "stop",
  "message": "OK"
}
```

If an error occurs:

```JSON
{
  "eventType": "stop",
  "error": true,
  "message": "Resource busy"
}
```

The `error` field must be set to `true` and the `message` field should contain a description of the error.

#### QUIT command

The `QUIT` command terminates the discovery. The response to `QUIT` is:

```JSON
{
  "eventType": "quit",
  "message": "OK"
}
```

after this output the discovery exits. This command is supposed to always succeed.

#### LIST command

The `LIST` command executes an enumeration of the ports and returns a list of the available ports at the moment of the
call. The format of the response is the following:

```
{
  "eventType": "list",
  "ports": [
    {
      "address":       <-- THE ADDRESS OF THE PORT
      "label":         <-- HOW THE PORT IS DISPLAYED ON THE GUI
      "protocol":      <-- THE PROTOCOL USED BY THE BOARD
      "protocolLabel": <-- HOW THE PROTOCOL IS DISPLAYED ON THE GUI
      "properties": {
                       <-- A LIST OF PROPERTIES OF THE PORT
      }

>     }, {

Reviewer: ```suggestion
    },
    {
```
The style used by all other JSON in the documentation.

      ...              <-- OTHER PORTS...
    }
  ]
}
```

The `ports` field contains a list of the available ports.

Each port has:

- an `address` (for example `/dev/ttyACM0` for serial ports or `192.168.10.100` for network ports)

- a `label` that is the human readable form of the `address` (it may be for example `ttyACM0` or
  `SSH on 192.168.10.100`)

- `protocol` is the protocol identifier (such as `serial` or `dfu` or `ssh`)

- `protocolLabel` is the `protocol` in human readable form (for example `Serial port` or `DFU USB` or `Network (ssh)`)

- `properties` is a list of key/value pairs that represent information relative to the specific port

> To make the above more clear let’s show an example with the `serial_discovery`:

Reviewer: I don't understand what `serial_discovery` is. I don't find any other occurrences of this identifier anywhere else in the documentation or the RFC.

Author: I changed it, should be clearer now.

    "protocolLabel": "Serial Port (USB)"
  }
}
```

It basically provides the same information as the `list` event but for a single port. After calling `START_SYNC` an
initial burst of add events must be generated in sequence to report all the ports available at the moment of the start.

The `remove` event looks like the following:

```JSON
{
  "eventType": "remove",
  "port": {
    "address": "/dev/ttyACM0",
    "protocol": "serial"
  }
}
```

The content is straightforward, in this case only the `address` and `protocol` fields are reported.

If the information about a port needs to be updated the discovery may send a new `add` message for the same port address
and protocol without sending a `remove` first: this means that all the previous information about the port must be
discarded and replaced with the new one.

#### Invalid commands

If the client sends an invalid or malformed command, the discovery should answer with:

```JSON
{
  "eventType": "command_error",
  "error": true,
  "message": "Unknown command XXXX"
}
```

### Board identification

The `properties` associated to a port can be used to identify the board attached to that port. The algorithm is simple:

> - each board listed in the platform file `boards.txt` may declare a set of `upload_port.*` properties

Reviewer: ```suggestion
- each board listed in the platform file [`boards.txt`](platform-specification.md#boardstxt) may declare a set of
  `upload_port.*` properties
```
The `upload_port.*` properties need to be documented in the platform specification.

Reviewer: @silvanocerza I see that my suggestion to linkify `boards.txt` was accepted, but I still didn't find any mention of the `upload_port` properties in the platform specification, so I think that has not been done.

Author: Yeah, completely forgot to add that on the platform specification.

Author: Added in https://github.com/arduino/arduino-cli/commit/efdb11c5cb7f7945e47a515f704830ec931f3fc5

Reviewer: Thanks!

- if each `upload_port.*` property has a match in the `properties` set coming from the discovery then the board is a

>   “candidate” board attached to that port.
> 
> Some port `properties` may not be precise enough to uniquely identify a board, in that case more boards may match the
> same set of `properties`, that’s why we called it “candidate”.
> 
> Let’s see an example to clarify things a bit, let's suppose that we have the following `properties` coming from the

Reviewer: ```suggestion
  "candidate" board attached to that port.

Some port `properties` may not be precise enough to uniquely identify a board, in that case more boards may match the
same set of `properties`, that's why we called it "candidate".

Let's see an example to clarify things a bit, let's suppose that we have the following `properties` coming from the
```
I don't trust this highfalutin "smart quotes" stuff and we don't use it in the rest of the documentation. It was introduced from Discourse, which forces them on you.

serial discovery:

> ```JSON
>   "port": {
>     "address": "/dev/ttyACM0",
>     "properties": {
>       "pid": "0x804e",
>       "vid": "0x2341",
>       "serialNumber": "EBEABFD6514D32364E202020FF10181E",
>       "name": "ttyACM0"
>     }

Reviewer: ````suggestion
```
  "port": {
    "address": "/dev/ttyACM0",
    "properties": {
      "pid": "0x804e",
      "vid": "0x2341",
      "serialNumber": "EBEABFD6514D32364E202020FF10181E",
      "name": "ttyACM0"
    },
    ...
````
Even though it is handled correctly by the Markdown renderers, the intentionally omitted missing closing brace in this code block really freaks out GitHub's diff viewer, which I think could cause confusion to future contributors.

```

in this case we can use `vid` and `pid` to identify the board. The `serialNumber`, instead, is unique for that specific

> instance of the board so it can't be used to identify the board model. Let’s suppose we have the following `boards.txt`:

Reviewer: ```suggestion
instance of the board so it can't be used to identify the board model. Let's suppose we have the following `boards.txt`:
```

mkr1000.upload_port.1.vid=0x2341
mkr1000.upload_port.1.pid=0x004e
mkr1000.upload_port.2.vid=0x2341
mkr1000.upload_port.2.pid=0x824e
mkr1000.upload_port.3.vid=0x2341
mkr1000.upload_port.3.pid=0x024e
[...CUT...]
```

As we can see the only board that has the two properties matching is the `mkr1000`, in this case the CLI knows that the
board is surely an MKR1000.

Note that `vid` and `pid` properties are just free text key/value pairs: the discovery may return basically anything,
the board just needs to have the same properties defined in `boards.txt` as `upload_port.*` to be identified.

We can also specify multiple identification properties for the same board using the `.N` suffix, for example:

```
myboard.name=My Wonderful Arduino Compatible Board
myboard.upload_port.pears=20
myboard.upload_port.apples=30
```

will match on `pears=20, apples=30` but:

```
myboard.name=My Wonderful Arduino Compatible Board
myboard.upload_port.0.pears=20
myboard.upload_port.0.apples=30
myboard.upload_port.1.pears=30
myboard.upload_port.1.apples=40
```

will match on both `pears=20, apples=30` and `pears=30, apples=40` but not `pears=20, apples=40`, in that sense each

> "set" of identification properties is indepentent from each other and cannot be mixed for port matching.

Reviewer: ```suggestion
"set" of identification properties is independent from each other and cannot be mixed for port matching.
```
